physx/src/foundation.rs

```diff
@@ -84,11 +93,33 @@ pub trait Foundation: Class<physx_sys::PxFoundation> + Sized {
     /// Returns `None` if `phys_PxCreateFoundation` returns a null pointer.
     fn new(allocator: Self::Allocator) -> Option<Owner<Self>> {
         unsafe {
-            Owner::from_raw(phys_PxCreateFoundation(
-                crate::physics::PX_PHYSICS_VERSION,
-                allocator.into_px(),
-                get_default_error_callback() as *mut PxErrorCallback,
-            ) as *mut Self)
+            Owner::from_raw(
+                phys_PxCreateFoundation(
+                    crate::physics::PX_PHYSICS_VERSION,
+                    allocator.into_px(),
+                    // needs to be create_default_error_callback rather than get_default_error_callback so we can delete it the same way as a custom callback
+                    create_default_error_callback().cast::<PxErrorCallback>(),
+                )
+                .cast::<Self>(),
+            )
+        }
+    }
+
+    /// Tries to create a PxFoundation with the provided allocator and error callbacks.
+    /// Returns `None` if `phys_PxCreateFoundation` returns a null pointer.
+    fn with_allocator_error_callback(
+        allocator: Self::Allocator,
+        error_callback: Box<dyn ErrorCallback>,
```

Comment:
I'm not super-enthused about this box/double-box thing we get going here. I realize the monomorphization is terrible for code-explosion, but this feels kludgy. Can we do something like convert it into a *PxErrorCallback in the builder already instead?

Reply:
I remember trying to do this using generics, but it got too complicated.
I can try again.

The double box though. Won't that still be needed at some level? Since our ffi is using void pointers for the user data?
Gonna have to fit that fat pointer into a usize somehow. Or is there some way to do vtables c++ style?
Perhaps I can make user data fit a fat pointer? But that wouldn't fit with our existing pattern either..

Reply:
With how the other trampolines are implemented (without box/dyn Trait), they don't have to pass a fat pointer - they end up passing a pointer to the concrete type. So we end up with a userdata +fn pointer trampoline. Pretty much a vtable.